### PR TITLE
schedulers: add macro support to metadata variables

### PR DESCRIPTION
**Pull request by jason-b-akers**
Summary:
This adds torchx.spec.macros support to the metadata variable values. This allows use cases like setting smcBridge to the app_id eg:
```
job_spec.roles[0].metadata["mast"] = {
        "HpcTaskGroupSpec": {
            "smcBridge": {
                "portName": "thrift",
                "smcTier": specs.macros.app_id,
            }
        }
    }
```

Reviewed By: manav-a

Differential Revision: D57519702

### CODE REVIEW

**Review by manav-a — APPROVED** [review]

File: torchx/specs/api.py [hunk 1]
```diff
@@ -189,8 +190,25 @@ def apply(self, role: "Role") -> "Role":
             role = copy.deepcopy(role)
             role.args = [self.substitute(arg) for arg in role.args]
             role.env = {key: self.substitute(arg) for key, arg in role.env.items()}
+            role.metadata = self._apply_nested(role.metadata)
```
Comment (kiukchung) [inline torchx/specs/api.py:193]: could you add a test-case for this?

Comment (manav-a) [reply torchx/specs/api.py:193]: There is a test case on the meta side of this make me can also add a version of it to api test @jason-b-akers

Comment (jason-b-akers) [reply torchx/specs/api.py:193]: Hi @kiukchung. There's an internal test in [D58678835](https://www.internalfb.com/diff/D58678835). As a follow-up I'll write one in torchx/schedulers/test/api_test.py as well.

Comment (jason-b-akers) [reply torchx/specs/api.py:193]: Follow-up test in [PR#293](github.com/pytorch/torchx/pull/923)

**Pull request merged**